narwhal/consensus/src/consensus.rs

```diff
@@ -31,6 +35,114 @@ pub mod consensus_tests;
 /// The representation of the DAG in memory.
 pub type Dag = BTreeMap<Round, HashMap<AuthorityIdentifier, (CertificateDigest, Certificate)>>;
 
+#[derive(Default)]
+pub struct LeaderSwapTable {
+    pub good_nodes: Vec<AuthorityIdentifier>,
+    pub bad_nodes: HashSet<AuthorityIdentifier>,
+}
+
+impl LeaderSwapTable {
+    // constructs a new table based on the provided reputation scores.
+    pub fn new(committee: &Committee, reputation_scores: ReputationScores) -> Self {
+        assert!(reputation_scores.final_of_schedule, "Only reputation scores that have been calculated on the end of a schedule are accepted");
+
+        let good_nodes = Self::good_nodes(&reputation_scores, committee);
+        let bad_nodes = Self::bad_nodes(&reputation_scores, committee);
+
+        debug!("Reputation scores: {:?}", reputation_scores);
+        debug!("Good nodes: {:?}", good_nodes);
+        debug!("Bad nodes: {:?}", bad_nodes);
+
+        Self {
+            good_nodes,
+            bad_nodes,
+        }
+    }
+
+    /// Checks whether the provided leader is a bad performer and needs to be swapped for the schedule
+    /// with a good performer. If not, then the method returns None. Otherwise the leader to swap with
+    /// is returned instead. The `leader_round` represents the DAG round on which the provided AuthorityIdentifier
+    /// is a leader on and is used as a seed to random function in order to calculate the good node that
```

Comment:
Based on the notion, I thought we were choosing the good node to swap in based on round robin fashion instead of random?

Reply:
You are absolutely right and I'll amend the Notion page to incorporate this approach - using randomness. After going into the implementation I figured out that following the round robin approach would be a bit more complex as we would need to keep more state to calculate the elected leaders on each round . Not impossible of course to implement, but more effort and potentially not much gain. Thus we switched to using seeded randomness instead so we can do things more stateless . I've coordinated with the Research team to ensure that we are not affecting other properties (ex fairness) and we seem to be good.